openeo/extra/job_management.py

```diff
@@ -0,0 +1,379 @@
+import collections
+import contextlib
+import datetime
+import json
+import logging
+import os
+import time
+from pathlib import Path
+from typing import Callable, Union, Dict, Optional
+
+import geopandas as gpd
+import pandas as pd
+import requests
+from openeo import Connection, BatchJob
+from openeo.rest import OpenEoApiError
+from openeo.util import deep_get
+
+from openeo.rest.connection import cached_connection
+
+
+_log = logging.getLogger(__name__)
+
+
+def run_jobs(
+    df: pd.DataFrame,
+    start_job: Callable,
+    outputFile: Path,
+    connection_provider: Callable,
+    parallel_jobs=2,
+):
+    """
+    Runs jobs, specified in a dataframe, and tracks parameters.
+
+    @param df: Job dataframe
+    @param start_job: A callback which will be invoked with the row of the dataframe for which a job should be started.
+    @param outputFile: A file on disk to track job statuses.
+    @return:
+    """
+
+    # TODO: original dataframe is completely discarded if `outputFile` exists, isn't that weird?
+    #       E.g. New code changes will not be picked up as long as an old/outdated CSV exist.
+    if outputFile.is_file():
+        df = pd.read_csv(outputFile)
+        df["geometry"] = gpd.GeoSeries.from_wkt(df["geometry"])
+    else:
+        df["status"] = "not_started"
+        df["start_time"] = ""
+        df["id"] = "None"
+        df["cpu"] = 0
+        df["memory"] = 0
+        df["duration"] = 0
+        df.to_csv(outputFile, index=False)
+
+    # TODO: this will never exit if there are failed/skipped jobs
+    while len(df[(df["status"] != "finished")]) > 0:
+        try:
+            jobs_to_run = df[df.status == "not_started"]
+            df = update_statuses(df, connection_provider)
+            df.to_csv(outputFile, index=False)
+            if jobs_to_run.empty:
+                time.sleep(60)
+                continue
+
+            if (
+                len(
+                    df[
+                        (df["status"] == "running")
+                        | (df["status"] == "queued")
+                        | (df["status"] == "created")
+                    ]
+                )
+                < parallel_jobs
+            ):
+                next_job = jobs_to_run.iloc[0]
+                job = start_job(next_job)
+                if job is not None:
+                    next_job["status"] = job.status()
+                    next_job["id"] = job.job_id
+                else:
+                    next_job["status"] = "skipped"
+                next_job["start_time"] = datetime.datetime.now().isoformat()
+                print(next_job)
+                df.loc[next_job.name] = next_job
+
+                df.to_csv(outputFile, index=False)
+            else:
+                time.sleep(60)
+
+        except requests.exceptions.ConnectionError as e:
+            _log.warning(f"Skipping connection error: {e}")
+
+
+def running_jobs(status_df):
```

Comment:
this method and the next one can also be removed, not used by job manager

Reply:
Has been removed.